DDD-1.md

```diff
@@ -0,0 +1,173 @@
+# DDD-1: Support for multiple connector tasks for relational Debezium connectors
+
+## Motivation
+
+Currently, it is challenging to use Debezium Connector for SQL Server in a multi-tenanted environment where a single SQL Server instance hosts tens to hundreds of similar tenant-scoped databases due to the fact that a given connector instance can capture changes only from a single database. The challenges are:
+
+1. Kafka Connect resources. Deploying hundreds of connector instances will require a Kafka Connect cluster with significant CPU and memory resources.
+2. Cluster stability. If the resources above are insufficient (e.g. memory), it may cause a Connect worker crash which in turn will fail all connector tasks scheduled to the worker. In the best case scenario, those tasks can be just restarted. In a worse case scenario, the restarted connectors will produce duplicates since the source connectors currently don't implement exactly-once processing semantics. In an even worse case, a worker can fail during a long-running snapshot, handling which will require additional cleanup.
+3. SQL Server resources. Each server instance will have to maintain an extra set of persistent connections (one per database, tens to hundreds in total) which may require additional server resources.
+
+## Proposed Changes
+
+### Source partition awareness at the task level ([#15](https://github.com/sugarcrm/debezium/pull/15))
+
+Currently, a Debezium connector task can process a single source partition throught its lifecycle. This partition is usually hard-coded in the context initialization code. E.g.:
+
+```java
+public SqlServerOffsetContext(SqlServerConnectorConfig connectorConfig) {
+        partition = Collections.singletonMap(SERVER_PARTITION_KEY, connectorConfig.getLogicalName());
```

Comment:
This will require some consideration when migrating existing connectors from the old offset format to the new one.

Reply:
Yes. I think it could be done in the form of a "legacy" switch like it's done for the MySQL connector but with an additional opt-in step:

1. The connector implements both the old and the new APIs.
2. A new configuration parameter is added to the connector (`internal.implementation=legacy|partitioned`). The default is `legacy`. Opt-in is possible.
3. In a subsequent release, the default is changed to `partitioned`, still possible to opt-out.
4. In another release, the configuration is removed together with the legacy implementation.

**Technical details of the migration**:

Whenever a user decides to upgrade an existing connector deployment:
1. Stop the old connector (one connector = one database).
2. Consume the message with the key `["$connector_name",{"server":"$connector_name"}]` from the Kafka Connect `offsets` topic.
3. Produce a message with the same value but `["$connector_name",{"server":"$connector_name", database: "$database_name"}]`.
4. Repeat for all connectors/databases you want to migrate under one new connector.
5. Start the new connector.

It looks like only the `offsets` topic will need care. The data produced by the connector remains the same.

Reply:
Yeah, perhaps it might even be easier. For other offset format changes we did, we just made sure that the new connector version can read back the old offset format. It'd always write the new one.